README
Welcome to pam_blue.

For install instruction see INSTALL - for further information 
consult -http://pam_blue.0xdef.net- (this link is dead, I'm not the
original author).

pam_blue is free software. Please see the file COPYING for details.
For documentation, please see the files in the doc subdirectory.
For building and installation instructions please see the INSTALL file.


NEWS
pam_blue -- History of visible changes.

Copyright (C) 2004, Hagen Paul Pfeifer <hagen@jauu.net>
See the end for copying conditions.

Please send blue_pam bug reports to hagen@jauu.net.

Version 0.1

* List your features here

-------------------------------------------------------
Copying information:

Copyright (C) 2004, Hagen Paul Pfeifer <hagen@jauu.net>

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.



src/utils.c
/*
** Copyright (C) 2004 Hagen Paul Pfeifer <hagen@jauu.net>
**   
** $Id: util.c 12 2004-04-21 14:02:44Z pfeifer $
**   
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software 
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#if HAVE_CONFIG_H
#include  <config.h>
#endif /* HAVE_CONFIG_H */

#include "global.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "utils.h"


#include <termios.h>

extern FILE *yyin;


Lst *lstptr;
struct entrybuff *ebp;


int gtimeout;

        
void
lst_init(Lst *lst, void (*destroy)(void *data))
{
  lst->size = 0;
  lst->destroy = destroy;
  lst->head = NULL;
  lst->tail = NULL;

  return;
}


void
lst_destroy(Lst *lst)
{
  void *data;

  while(lst_size(lst) > 0) {
    if (lst_rem_nxt(lst, NULL, (void **)&data) == 0 && lst->destroy != NULL) {
      lst->destroy(data);
    }
  }
  
  memset(lst, 0, sizeof(Lst));
  return;
}


int
lst_ins_nxt(Lst *lst, Lst_mbr *lst_mbr, const void *data)
{
  Lst_mbr *new_mbr;

  if ((new_mbr = (Lst_mbr *) malloc(sizeof(Lst_mbr))) == NULL) {
    return -1;
  }
  new_mbr->data = (void *) data;
  if (lst_mbr == NULL) {
    if (lst_size(lst) == 0) {
      lst->tail = new_mbr;
    }
    
    new_mbr->next = lst->head;
    lst->head = new_mbr;

  } else {
    
    if (lst_mbr->next == NULL) {
      lst->tail = new_mbr;
    }

    new_mbr->next = lst_mbr->next;
    lst_mbr->next = new_mbr;
  }

  ++lst->size;
  return 0;
}



int
lst_rem_nxt(Lst *lst, Lst_mbr *lst_mbr, void **data)
{

  Lst_mbr *old_mbr;

  if (lst_size(lst) == 0) {
    return -1;
  }

  if (lst_mbr == NULL) {
    *data   = lst->head->data;
    old_mbr = lst->head;
    lst->head = lst->head->next;
    if (lst_size(lst) == 1)

      lst->tail = NULL;



  } else {

    if (lst_mbr->next == NULL) {
      return -1;
    }

    *data = lst_mbr->next->data;
    old_mbr = lst_mbr->next;
    lst_mbr->next = lst_mbr->next->next;

    if (lst_mbr->next == NULL) {
      lst->tail = lst_mbr;
    }

  }

  free(old_mbr);
  --lst->size;

  return 0;
  
}


/* Open configuration file and iterate over all entrys in it
 * malloc for every entry a entrybuff struct and add to list
 * 
 */
Lst *
parse_rc()
{
  
  /* FIXME: hard coded path */
  if ( (yyin = fopen(CONFIG_FILE, "r")) == NULL) {
      syslog (LOG_ERR, "Can't find configuration file (" CONFIG_FILE ")!");
      return NULL;
  }

  if ((lstptr = (Lst *) malloc(sizeof(Lst))) == NULL) {
    syslog(LOG_ERR, "Module Error! (Can't malloc(2) memory)");
    return NULL;
  }
  
  if ((ebp = (struct entrybuff *) malloc(sizeof(struct entrybuff))) == NULL) {
    syslog(LOG_ERR, "Module Error! (Can't malloc(2) memory)");
    return NULL;
  }

  /* bzero(3) struct for parser */
  init_struct(ebp);

  /* create list */
  lst_init(lstptr, (void *) clear_struct);
 
 if (yyparse())
   return NULL;

 /* print_list(); */


 return lstptr;

}

int
clear_struct(struct entrybuff *eb)
{

  if (eb->name != NULL) {
    free(eb->name);
  }
  
  if (eb->btmac != NULL) {
    free(eb->btmac);
  }

  if (eb->btname != NULL) {
    free(eb->btname);
  }

  if (eb) {
    free(eb);
    eb = NULL;
  }
  
  return 0;
}

int
init_struct(struct entrybuff *eb)
{
  eb->name    = NULL;
  eb->btmac   = NULL;
  eb->btname  = NULL;
  eb->timeout = 0;
  return 0;
}

int
set_gtimeout(int timeout)
{
  gtimeout = timeout;
}

int
set_userentry(struct entrybuff *eb)
{

  lst_ins_nxt(lstptr, lst_tail(lstptr), eb);

  if (eb->timeout == 0) { /* take default one */
    eb->timeout = gtimeout;
  }
  
  if ((ebp = (struct entrybuff *) malloc(sizeof(struct entrybuff))) == NULL) {
    syslog(LOG_ERR, "Module Error! (Can't malloc(2) memory)");
    exit(1);
  }

  /* bzero(3) struct for parser */
  init_struct(ebp);
  
  return 0;

}

#if 0  /* module debug entry */
int
print_list(void)
{
  int i;

  Lst_mbr *lst_mbrp = lstptr->head;
  for(i = 0; i < lst_size(lstptr); i++ ) {

    struct entrybuff *bptr = (struct entrybuff *) lst_mbrp->data;
    printf("\nUSERNAME := %s\n", bptr->name);
    printf("btmac  : %s ::\n", bptr->btmac);
    printf("btname : %s ::\n", bptr->btname);
    printf("timeout: %d ::\n", bptr->timeout);

    lst_mbrp = lst_mbrp->next;

  }

  return 0;

}

#endif  /* module debug entry */



src/utils.h
/*
** Copyright (C) 2003 Hagen Paul Pfeifer <hagen@jauu.net>
**  
** $Id: util.h 13 2004-04-21 17:07:16Z pfeifer $
**  
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software 
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#ifndef UTILS_H_DEF
#define UTILS_H_DEF

#include <stdlib.h>


#define USER  1
#define GROUP 2

/* elements of Lst */
typedef struct _lst_mbr {
    
    void            *data;
    struct _lst_mbr *next;

} Lst_mbr;

/* nice generic list ;-) */
typedef struct _lst {

    int  size;
    int  (*match)(const void *key1, const void *key2);
    void (*destroy)(void *data);
    
    Lst_mbr *head;
    /* maybe needed */
    Lst_mbr *tail;

} Lst;

/* struct for parser run */
struct entrybuff {
    int  type;
    char *name;
    char *btmac;
    char *btname;
    int  timeout;
};

/* public prototypes */
void lst_init(Lst *, void (*destroy)(void *data));
void lst_destroy(Lst *);
int  lst_ins_nxt(Lst *, Lst_mbr *, const void *);
int  lst_rem_nxt(Lst *, Lst_mbr *, void **);

int clear_struct(struct entrybuff *);
int init_struct(struct entrybuff *);
int set_userentry(struct entrybuff *);
int print_list(void);

#define lst_size(__lst) ((__lst)->size)
#define lst_head(__lst) ((__lst)->head)
#define lst_tail(__lst) ((__lst)->tail)

#define data_is_head(__lst, __lst_mbr) ((__lst_mbr) == (__lst)->head ? 1 : 0)
#define data_is_tail(__lst_mbr) ((__lst_mbr)->next == NULL ? 1 : 0)

#define lst_data(__lst_mbr) ((__lst_mbr)->data)
#define lst_next(__lst_mbr) ((__lst_mbr)->next)


#endif /* UTILS_H_DEF */


src/global.h
/*
** Copyright (C) 2003 Hagen Paul Pfeifer <hagen@jauu.net>
**  
** $Id: global.h 13 2004-04-21 17:07:16Z pfeifer $
**  
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software 
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#ifndef GLOBAL_H_DEF
#define GLOBAL_H_DEF

#include <syslog.h>

#define CONFIG_FILE "/etc/security/bluesscan.conf"
#define DEFAULTTIMEOUT 5

#define MAXLINE 1024

#include "utils.h"

#ifdef __GCC__
#define max(a, b) \
    ({typedef _ta = (a), _tb = (b); \
     _ta _a = (a); _tb _b = (b);    \
     _a > _b ? _a : _b; })
#else
#define max(a, b) \
    ((a < b) ? (a) : (b))
#endif /* __GCC */


/* bluescan.c */
int bluescan(struct entrybuff *enbfptr);
Lst *parse_rc(void);
#endif /* GLOBAL_H_DEF */


src/bluescan.c
/*
** Copyright (C) 2003 Hagen Paul Pfeifer <hagen@jauu.net>
**
** $Id: bluescan.c 19 2004-04-21 18:24:46Z pfeifer $
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#if HAVE_CONFIG_H
#include  <config.h>
#endif /* HAVE_CONFIG_H */

#include "global.h"


#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <fcntl.h>
#include <errno.h>
#include <ctype.h>

#include <termios.h>
#include <fcntl.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/poll.h>
#include <netinet/in.h>

#include <bluetooth/bluetooth.h>
#include <bluetooth/hci.h>
#include <bluetooth/hci_lib.h>
#include <bluetooth/l2cap.h>

#define	false		-1
#define	true		1

#ifdef DEBUG
static inline void
show_debug(char *msg)
{
	fprintf(stderr,"bluescan: %s\n",msg);
}
#else
static inline void show_debug(char *msg) {}
#endif /* DEBUG */

/*
** derived form bluez-utils/tools/l2ping.c
** tim - tim@spacetim.de
*/

/* Improved connection timeout 
** David De Sousa - davidesousa@gmail.com
*/

int bluescan
(struct entrybuff *enbfptr)
{
	struct sockaddr_l2 addr;
	socklen_t optlen;
	unsigned char *buf;
	char str[18];
	int i, sk, lost;
	int size    = 44;
	int ident   = 200;
	int timeout;
	bdaddr_t bdaddr;
	long skctl; // Socket Flags
	/* Setting variables for select to work. */
	fd_set skset; 
	struct timeval tv;
	int lon, retval;

	bacpy(&bdaddr, BDADDR_ANY);

	if ( (enbfptr->timeout < 3) || (enbfptr->timeout > 15))
		timeout = DEFAULTTIMEOUT;
	else
		timeout = enbfptr->timeout;


	buf = malloc(L2CAP_CMD_HDR_SIZE + size);
	if (!buf) {
		show_debug("Can't allocate buffer");
		return(false);
	}

	/* Create socket */
	sk = socket(PF_BLUETOOTH, SOCK_RAW, BTPROTO_L2CAP);
	if (sk < 0) {
		show_debug("Can't create socket");
		free(buf);
		return(false);
	}

	/* Bind to local address */
	memset(&addr, 0, sizeof(addr));
	addr.l2_family = AF_BLUETOOTH;
	bacpy(&addr.l2_bdaddr, &bdaddr);

	if (bind(sk, (struct sockaddr *) &addr, sizeof(addr)) < 0) {
		show_debug("Can't bind socket");
		close(sk);
		free(buf);
		return(false);
	}

	/* Set non-blocking state for the socket. DDS */
        skctl = fcntl(sk, F_GETFL, NULL);
        skctl |= O_NONBLOCK;
        fcntl(sk, F_SETFL, skctl);
				

	/* Connect to remote device */
	memset(&addr, 0, sizeof(addr));
	addr.l2_family = AF_BLUETOOTH;
	str2ba(enbfptr->btmac, &addr.l2_bdaddr);

	/* Connecting with timeout. */
	if (connect(sk, (struct sockaddr *) &addr, sizeof(addr)) < 0) {
		if (errno == EINPROGRESS) {
			tv.tv_sec = timeout; 
		        tv.tv_usec = 0;
			FD_ZERO(&skset); 
		        FD_SET(sk, &skset);
			if (select(sk+1, NULL, &skset, NULL, &tv) > 0) { 
				lon = sizeof(int); 
				getsockopt(sk, SOL_SOCKET, SO_ERROR, (void*)(&retval), &lon); 
				if (retval)
					show_debug ("Error Connecting!!!");
				else
					show_debug ("Connected!!!");
        		} 
			else {
				show_debug ("Connection Timeout [bluetooth device up?]");
				retval = 1;
			}
		}
		else {
			show_debug ("Error Connecting");
			retval = 1;
		}
		
		if (retval) {
	                close(sk);
        	        free(buf);
                	return(false);
		}
	}

	/* Set socket to original mode (blocking). */

	skctl = fcntl(sk, F_GETFL, NULL); 
	skctl &= (~O_NONBLOCK); 
	fcntl(sk, F_SETFL, skctl); 

	/* Get local address */
	memset(&addr, 0, sizeof(addr));
	optlen = sizeof(addr);

	if (getsockname(sk, (struct sockaddr *) &addr, &optlen) < 0) {
		show_debug("Can't get local address");
		close(sk);
		free(buf);
		return(false);
	}

	ba2str(&addr.l2_bdaddr, str);

	/* Initialize buffer */
	for (i = 0; i < size; i++)
		buf[L2CAP_CMD_HDR_SIZE + i] = (i % 40) + 'A';

	l2cap_cmd_hdr *cmd = (l2cap_cmd_hdr *) buf;

	/* Build command header */
	cmd->code  = L2CAP_ECHO_REQ;
	cmd->ident = ident;
	cmd->len   = htobs(size);

	/* Send Echo Request */
	if (send(sk, buf, L2CAP_CMD_HDR_SIZE + size, 0) <= 0) {
		show_debug("Send failed");
		close(sk);
		free(buf);
		return(false);
	}

	/* Wait for Echo Response */
	lost = 0;
	while (1) {
		struct pollfd pf[1];
		int err;

		pf[0].fd = sk;
		pf[0].events = POLLIN;

		if ((err = poll(pf, 1, timeout * 1000)) < 0) {
			show_debug("Poll failed");
			close(sk);
			free(buf);
			return(false);
		}

		if (!err) {
			lost = 1;
			break;
		}

		if ((err = recv(sk, buf, L2CAP_CMD_HDR_SIZE + size, 0)) < 0) {
			show_debug("Recv failed");
			close(sk);
			free(buf);
			return(false);
		}

		if (!err){
			show_debug("Disconnected\n");
			close(sk);
			free(buf);
			return(false);
		}

		cmd->len = btohs(cmd->len);

		if (cmd->ident != ident)
			continue;

		if (cmd->code == L2CAP_ECHO_RSP)
			break;
		if (cmd->code == L2CAP_COMMAND_REJ) {
			show_debug("Peer doesn't support Echo packets\n");
			close(sk);
			free(buf);
			return(false);
		}

	}
	show_debug("pam_blue successful!");
	close(sk);
	free(buf);
	return(true);
}


src/bluepmain.c
/*
** Copyright (C) 2003 Hagen Paul Pfeifer <hagen@jauu.net>
**  
** $Id: bluepmain.c 19 2004-04-21 18:24:46Z pfeifer $
**  
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software 
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#if HAVE_CONFIG_H
#include  <config.h>
#endif /* HAVE_CONFIG_H */

#include "global.h"

#define DEFAULT_USER "nobody"

#include <stdio.h>

#define PAM_SM_AUTH
#define PAM_SM_ACCOUNT
#define PAM_SM_SESSION
#define PAM_SM_PASSWORD

#include <security/pam_modules.h>
#include <security/_pam_macros.h>

#include <sys/types.h>
#include <pwd.h>

#define _XOPEN_SOURCE
#include <unistd.h>

int debug = 0;
 
PAM_EXTERN int
pam_sm_authenticate(pam_handle_t *pamh, int flags,
                    int argc, const char **argv)
{
  int i, error = 0;
  int retval;
  char *reterror = NULL;
  struct pam_conv *conv;
  struct pam_message  *pmsg[3], msg[3];
  struct pam_response *response;   
  struct entrybuff *bptr;
  const char *user=NULL;
  Lst *lst;
  Lst_mbr *lst_mbrp;

  openlog("pam_blue", LOG_NDELAY | LOG_CONS | LOG_PID, LOG_AUTHPRIV);

  if (argc >= 1)
    if (!strncmp(argv[0], "debug", 5))
      debug = 1;
  
  /* whoWeAre.com ;-) */
  retval = pam_get_user( pamh, &user, NULL);
  if (retval != PAM_SUCCESS) {
    syslog (LOG_ERR, "Can't determine username");
    return PAM_USER_UNKNOWN;
  }

  retval = pam_get_item(pamh, PAM_CONV, (const void **) &conv);
  if (retval != PAM_SUCCESS) {
    user = NULL;
    syslog (LOG_ERR, "pam errror [%s]!", (pam_strerror (pamh,retval)));
    return PAM_SYSTEM_ERR;
  } 
  
  if (user == NULL || *user == '\0' ) {
    /* D (("pam_get_user: invalid username")); */
    user = NULL;
    syslog (LOG_ERR, "Invalid username!");
    return PAM_USER_UNKNOWN;
  }

  if ( (lst = parse_rc()) == NULL) { /* something goes wrong (no config file, ...) */
    user = NULL;
    syslog (LOG_ERR, "Can't parse configuration file [%s]!" CONFIG_FILE);
    return PAM_AUTH_ERR;
  }
    
  /* fine we had a valid username; check for user in list
   * break if we found the first valid entry
   * TODO: implement something like priorities: user are higher rated then
   * group entrys
   * 
   * Starting at this point make sure you free(3) all malloced datastructs!
   * 
   */
  
  lst_mbrp = lst->head;
  reterror = "User not found in configuration file [" CONFIG_FILE "]!";
  for(i = 0; i < lst_size(lst); i++) {
    bptr = (struct entrybuff *) lst_mbrp->data;
    if (! strcmp(bptr->name, user)) { /* found user in list */
      reterror = "User found in configuration file [" CONFIG_FILE "], but scan failed!";
      /* TODO: add some error return codes for speparate logging */
      if (bluescan(bptr) > 0) {
        if (debug)
          syslog (LOG_ERR, "scan succeed for user %s!", user);
        user = NULL;
        lst_destroy(lst);
        return PAM_SUCCESS;
      } else { /* failure */
        syslog (LOG_ERR, "Bluetooth scan failure [bluetooth device %s up?]", bptr->btmac);
      }
    }
    lst_mbrp = lst_mbrp->next;
  }

  if (reterror) { /* free struct, log and return failure */
    user = NULL;
    lst_destroy(lst);
    syslog (LOG_ERR, "%s", reterror);
    return PAM_AUTH_ERR;

  }

  /* catch all rule */
  user = NULL;
  lst_destroy(lst);
  syslog (LOG_ERR, "Programmed failure!!!");
  return PAM_AUTH_ERR;
}

PAM_EXTERN
int pam_sm_setcred(pam_handle_t *pamh,int flags,int argc
    ,const char **argv)
{
  return PAM_SUCCESS;
}

/* --- account management functions --- */

PAM_EXTERN
int pam_sm_acct_mgmt(pam_handle_t *pamh,int flags,int argc
    ,const char **argv)
{
  return PAM_SUCCESS;
}

/* --- password management --- */

PAM_EXTERN
int pam_sm_chauthtok(pam_handle_t *pamh,int flags,int argc
    ,const char **argv)
{
  return PAM_SUCCESS;
}

/* --- session management --- */

PAM_EXTERN
int pam_sm_open_session(pam_handle_t *pamh,int flags,int argc
    ,const char **argv)
{
  return PAM_SUCCESS;
}

PAM_EXTERN
int pam_sm_close_session(pam_handle_t *pamh,int flags,int argc
    ,const char **argv)
{
  return PAM_SUCCESS;
}

/* end of module definition */

#ifdef PAM_STATIC

/* static module data */

struct pam_module _pam_permit_modstruct = {
  "pam_permit",
  pam_sm_authenticate,
  pam_sm_setcred,
  pam_sm_acct_mgmt,
  pam_sm_open_session,
  pam_sm_close_session,
  pam_sm_chauthtok
};

#endif

#if 0
  /* searching prompt */
  if (1)
  {
    pmsg[0] = &msg[0];
    msg[0].msg_style = PAM_TEXT_INFO;
    msg[0].msg = (char *) malloc (sizeof (char) * (50 + strlen (user)));
    snprintf (msg[0].msg, 50 + strlen (user), "Scanning ... ");
    retval = conv->conv (1, (const struct pam_message **) pmsg, &response, conv->appdata_ptr);
  }

  
  /* ..and found prompt */
  if (1)
  {
    pmsg[0] = &msg[0];
    msg[0].msg_style = PAM_TEXT_INFO;
    msg[0].msg = (char *) malloc (sizeof (char) * (50 + strlen (user)));
    snprintf (msg[0].msg, 50 + strlen (user), "found");
    retval = conv->conv (1, (const struct pam_message **) pmsg, &response, conv->appdata_ptr);
  }

#endif 
